src/HexagonOptimize.cpp

```diff
@@ -414,9 +455,47 @@ class OptimizePatterns : public IRMutator {
     }
 
     void visit(const Cast *op) {
+        // To hit more of the patterns we want, rewrite "double casts"
+        // as two stage casts. This also avoids letting vector casts
+        // fall through to LLVM, which will generate large unoptimized
+        // shuffles.
+        static vector<pair<Expr, Expr>> cast_rewrites = {
+            // Saturating narrowing
+            { u8c(wild_u32x), u8c(u16c(wild_u32x)) },
+            { u8c(wild_i32x), u8c(i16c(wild_i32x)) },
+            { i8c(wild_u32x), i8c(u16c(wild_u32x)) },
+            { i8c(wild_i32x), i8c(i16c(wild_i32x)) },
+
+            // Narrowing
+            { u8(wild_u32x), u8(u16(wild_u32x)) },
+            { u8(wild_i32x), u8(i16(wild_i32x)) },
+            { i8(wild_u32x), i8(u16(wild_u32x)) },
+            { i8(wild_i32x), i8(i16(wild_i32x)) },
+
+            // Widening
+            { u32(wild_u8x), u32(u16(wild_u8x)) },
+            { u32(wild_i8x), u32(i16(wild_i8x)) },
```

Comment:
I just noticed a bug here, this should widen to u16, not i16, otherwise it will do a weird half-way sign extension.